docs/lsp.md

```diff
@@ -0,0 +1,489 @@
+
+We need to investigate which system we'll use for the New Vsix. Ideally we're able to go with new VisualStudio.Extensibility but if that has any blockers we might need to stick to in-process MEF.
+
+We can prototype the New Vsix with some sample LSP server or maybe with an existing one like FSAC. Just to figure out how to connect it to F# projects in VS.
+
+For FsLSP we need to choose a library and implement a dummy LSP server with it that we can connect to the New Vsix and get some sort of basic setup that can then be improved upon.
+
+<!-- Note: these charts don't contain actual time estimates -->
+
+```mermaid
+
+gantt
+    axisFormat .
+    section FsLSP
+        Choose LSP library :fs1, 1900-01-01, 1d
+        Add projects for FsLSP and FsLSPServer :fs2, after fs1, 1d
+        Implement dummy LSP server using the library :fs3, after fs2, 1d
+        Connect to NewVsix :after vsix2 fs3, 1d
+    section Vsix
+        Investigate options for NewVsix :vsix1, 1900-01-01, 1d
+        Implement NewVsix prototype with dummy or some existing LSP server :vsix2, after vsix1, 2d
+```
+
+### Phase 1
+
+Get some sort of working prototype that can do semantic highlighting and show diagnostics.
+
+Figure out how to run side-by-side and gradually take over functionality from the current Vsix, if possible.
+
+```mermaid
+
+gantt
+    axisFormat .
+    section FsLSP
+        Implement WorkspaceStateManagement :fs1, 1900-01-01, 3d
+        Implement Phase 1 endpoints :fs2, after fs1, 2d
+    section Vsix
+        Investigate running side-by-side with old Vsix :vsix1, 1900-01-01, 1d
+        Implement configuration of which features are handled by which Vsix :vsix2, after vsix1, 2d
+```
+
+### Phase 2 - 5
+
+Implement endpoints based on their assigned priority. See next section.
+
+Independently we can start on setting up the build & release process for FsLSP and insertions of NewVsix into VS.
+
+Another potentially parallel track can be figuring out interop with C# LSP server for mixed solutions.
+
+```mermaid
+
+gantt
+    axisFormat .
+    section FsLSP
+        Implement Phase 2 endpoints :fs1, 1900-01-01, 1d
+        Implement Phase 3 endpoints :fs2, after fs1, 1d
+        Implement Phase 4 endpoints :fs3, after fs2, 1d
+        Implement Phase 5 endpoints :fs4, after fs3, 1d
+    section Distribution
+        Settle on release process :d1, 1900-01-01, 1d
+        Add / update build pipelines :d2, after d1, 1d
+        Set up VS insertions :d3, after d2, 1d
+
+    section Cs interop
+        Figure out how to talk to CSharp LSP server :c1, 1900-01-01, 4d
+```
+
+
+
+## LSP Endpoints
+
+This is a preliminary list of LSP APIs we want (or don't want) to implement, roughly ordered by priority. We will probably re-classify some of them as we go.
+
+
+| Name                                     | Type         | Direction | Category                          | Priority / Dev phase | Supported by FSAC | Note                                              |
```

Comment:
@0101
Based on yesterday's meeting with Cyrus, David and Todd, shall we add a columt which would mark whether certain requests supported by default VS client?

Reply:
Yeah, I'll try to find out which ones they are. 

But I'm suspecting we won't be able to function just with whatever VS does by itself and we will have to add a custom client.

Reply:
I think we should be good with most (if not all) **core** ones, since there are working LSP servers in VS right now, but might need to "mock" the client with MEF or ask LSP team to implement missing functionality.